src/core/public/application/application_service.tsx

```diff
@@ -17,159 +17,154 @@
+        const handler = this.mountContext!.createHandler(plugin, app.mount);
+        const mount: AppMounter = async params => {
+          const unmount = await handler(params);
+          this.currentAppId$!.next(app.id);
+          return unmount;
         };
-        this.apps$.next(new Map([...this.apps$.value.entries(), [app.id, appBox]]));
+        this.apps.set(app.id, app);
+        this.mounters.set(app.id, { appBasePath: app.appBasePath!, mount });
       },
-      registerLegacyApp: (app: LegacyApp) => {
-        if (this.legacyApps$.value.has(app.id)) {
-          throw new Error(`A legacy application is already registered with the id "${app.id}"`);
-        }
-        if (this.legacyApps$.isStopped) {
+      registerLegacyApp: app => {
+        const appBasePath = `/app/${app.id.split(':')[0]}`;
+
+        if (this.registrationClosed) {
           throw new Error(`Applications cannot be registered after "setup"`);
+        } else if (this.legacyApps.has(app.id)) {
+          throw new Error(`A legacy application is already registered with the id "${app.id}"`);
+        } else if (hasAppBasePath(this.mounters, appBasePath)) {
+          throw new Error(
+            `An application is already registered with the appBasePath "${appBasePath}"`
+          );
         }
 
-        this.legacyApps$.next(new Map([...this.legacyApps$.value.entries(), [app.id, app]]));
+        const mount: LegacyAppMounter = () => redirectTo(basePath.prepend(appBasePath));
+        this.legacyApps.set(app.id, app);
+        this.mounters.set(app.id, { appBasePath, mount, unmountBeforeMounting: true });
       },
-      registerMountContext: this.mountContext!.registerContext,
     };
   }
 
-  public async start({
-    http,
-    injectedMetadata,
-    redirectTo = (path: string) => (window.location.href = path),
-  }: StartDeps): Promise<InternalApplicationStart> {
+  public async start({ injectedMetadata }: StartDeps): Promise<InternalApplicationStart> {
     if (!this.mountContext) {
-      throw new Error(`ApplicationService#setup() must be invoked before start.`);
+      throw new Error('ApplicationService#setup() must be invoked before start.');
     }
 
-    // Disable registration of new applications
-    this.apps$.complete();
-    this.legacyApps$.complete();
-
-    const legacyMode = injectedMetadata.getLegacyMode();
-    const currentAppId$ = new BehaviorSubject<string | undefined>(undefined);
-    const { availableApps, availableLegacyApps, capabilities } = await this.capabilities.start({
-      apps: new Map([...this.apps$.value].map(([id, { app }]) => [id, app])),
-      legacyApps: this.legacyApps$.value,
-      injectedMetadata,
-    });
-
-    // Only setup history if we're not in legacy mode
-    const history = legacyMode ? null : createBrowserHistory({ basename: http.basePath.get() });
+    this.registrationClosed = true;
+    this.currentAppId$ = new BehaviorSubject<string | undefined>(undefined);
```

Comment:
Shouldn't `currentAppId$` be completed on stop?

Reply:
The application service is never stopped.

Reply:
Could you elaborate on what makes it so special? AFAIK the platform stopped all other services. https://github.com/restrry/kibana/blob/2f87fd624aab8c3c5fbe65094ae2a72e38dcb51e/src/core/public/core_system.ts#L300-L309

Reply:
In the current implementation the application service is never stopped and this observable is never completed. @joshdover thoughts on stopping the application service? I thought it was redundant since Kibana can't really run without it, but maybe this isn't the case?

Reply:
I believe all services should be stopped when the FatalErrorService is invoked. I don't see any reason why the ApplicationService couldn't be stopped then (though I see that right now it's not included in `CoreSystem#stop()`, that seems like an inadvertant omission).